metadata-ingestion/src/datahub/utilities/memory_leak_detector.py

```diff
@@ -0,0 +1,123 @@
+    unique_traces_to_objects: Dict[
+        Union[tracemalloc.Traceback, int], List[object]
+    ] = defaultdict(list)
+    for obj in gc.garbage:
+        obj_trace = tracemalloc.get_object_traceback(obj)
+        if obj_trace is not None:
+            if _trace_has_file(obj_trace, "*datahub/*.py"):
+                # Leaking object
+                unique_traces_to_objects[obj_trace].append(obj)
+            else:
+                unique_traces_to_objects[id(obj)].append(obj)
+    logger.info("Potentially leaking objects start")
+    for key, obj_list in sorted(
+        unique_traces_to_objects.items(),
+        key=lambda item: sum([sys.getsizeof(o) for o in item[1]]),
+        reverse=True,
+    ):
+        if isinstance(key, tracemalloc.Traceback):
+            obj_traceback: tracemalloc.Traceback = cast(tracemalloc.Traceback, key)
+            logger.info(
+                f"#Objects:{len(obj_list)}; Total memory:{sum([sys.getsizeof(obj) for obj in obj_list])};"
+                + " Allocation Trace:\n\t"
+                + "\n\t".join(obj_traceback.format(limit=25, most_recent_first=True))
+            )
+        else:
+            logger.info(
+                f"#Objects:{len(obj_list)}; Total memory:{sum([sys.getsizeof(obj) for obj in obj_list])};"
+                + " No Allocation Trace available!"
+            )
+        # Print details about the live referrers of each object in the obj_list (same trace).
+        for obj in obj_list:
+            referrers = [r for r in gc.get_referrers(obj) if r in gc.garbage]
+            logger.info(
+                f"Referrers[{len(referrers)}] for object(addr={hex(id(obj))}):'{obj}':"
+            )
+            for ref_index, referrer in enumerate(referrers):
+                ref_trace = tracemalloc.get_object_traceback(referrer)
+                logger.info(
+                    f"Referrer[{ref_index}] referrer_obj(addr={hex(id(referrer))}):{referrer}, RefTrace:\n\t\t"
+                    + "\n\t\t".join(
+                        ref_trace.format(limit=5, most_recent_first=True)
+                        if ref_trace
+                        else []
+                    )
+                    + "\n"
+                )
+    logger.info("Potentially leaking objects end")
+
+    tracemalloc.stop()
+
+
+def with_leak_detection(func: Callable[..., T]) -> Callable[..., T]:
+    @wraps(func)
+    def wrapper(*args: Any, **kwargs: Any) -> Any:
+        detect_leaks: bool = args[0].obj.get("detect_memory_leaks", False)
+        exception: Optional[BaseException] = None
+        if detect_leaks:
+            logger.info(
+                f"Initializing memory leak detection on command: {func.__module__}.{func.__name__}"
+            )
+            _init_leak_detection()
+
+        try:
+            res = func(*args, **kwargs)
+        # Catch all exceptions, including SystemExit that some of the commands raise via sys.exit()
+        except BaseException as e:
+            exception = e
```

Comment:
Essentially the code after line 108 is technically `finally` here. I still need to catch and rethrow because we can't process SystemExit exception before performing the leak detection. Will change the code to explicitly reflect that.

Reply:
> I still need to catch and rethrow because we can't process SystemExit exception before performing the leak detection

What do you mean by that? Does the triggering of a SystemExit mess with the leak detection in some way?

Reply:
Yup, try...finally is the right thing to use here. Thanks for calling out :-)